### PR TITLE
Preview 2 filesystem: track open_mode instead of reporting the permissions

### PR DESCRIPTION
**Pull request by pchickey**
First update the fd_filestat_set test to assert that the filestat_set operations on a file descriptor behaves the same whether opened readonly or readwrite. This test now reproduces the behavior reported in https://github.com/bytecodealliance/wasmtime/issues/7829. wasi-common passes this changed test, but wasmtime-wasi::preview2 does not, showing that this is a regression between the two implementations.

Then, in preview2 implementation, track open_mode for a File and Dir, and use that instead of reporting permissions in DescriptorFlags.

FilePerms/DirPerms is a way of having wasmtime-wasi, instead of the operating system, mediate permissions on a preview2 Descriptor. This was conflated with the open mode, which should determine whether DescriptorFlags contains READ or WRITE or MUTATE_DIRECTORY.

* don't mask FilePerms with the DescriptorFlags (oflags) in path_open
* File and Dir now track their open_mode (represented using FilePerms)
  and report them in DescriptorFlags

### CODE REVIEW

File: crates/test-programs/src/bin/preview1_fd_filestat_set.rs [hunk 1]
```diff
@@ -81,5 +73,11 @@ fn main() {
     };
 
     // Run the tests.
-    unsafe { test_fd_filestat_set(dir_fd) }
+    //
+    // The filestat_set functions should behave the same whether the file
+    // descriptor is open for read-only or read-write, because the underlying
+    // permissions of the file determine whether or not the filestat can be
+    // set or not, not than the open mode.
+    unsafe { test_fd_filestat_set(dir_fd, wasi::RIGHTS_FD_READ) }
```
Comment (sunfishcode) [inline crates/test-programs/src/bin/preview1_fd_filestat_set.rs:81]: This doesn't seem right. If we open the file with just `RIGHTS_FD_READ`, we shuldn't be able to do `fd_filestat_set_size` on it. That would be analogous in POSIX to opening a file with `FD_READ` and calling `ftruncate`, which is defined to fail.

Comment (pchickey) [reply crates/test-programs/src/bin/preview1_fd_filestat_set.rs:81]: That was what I thought too but I didnt chase it down in the ftruncate case. From the python test suite I found that futime is expected to succeed even when a file is read-only. So, I'll have to change this test, as well as set-len to check the open mode and reject size changes when not open for writing (fail with EINVAL, which I prefer here to EBADF) but permit set-times.

Comment (pchickey) [reply crates/test-programs/src/bin/preview1_fd_filestat_set.rs]: This problem ended up being down to how the test was written. wasi::OFLAGS_CREAT implies that the WASI impl open the file with the O_RRONLY or O_RDWR in posix, so we were never actually opening a file just for reading in this test. I changed the test to create the file, then open it under the requested mode, and it behaves as expected.

**Comment by pchickey** [comment]
I believe the test now shows that the implementations all behave as expected.

File: crates/wasi/src/preview2/ctx.rs [hunk 2]
```diff
@@ -144,8 +144,17 @@ impl WasiCtxBuilder {
         file_perms: FilePerms,
         path: impl AsRef<str>,
     ) -> &mut Self {
-        self.preopens
-            .push((Dir::new(dir, perms, file_perms), path.as_ref().to_owned()));
+        let mut open_flags = FilePerms::empty();
```
Comment (alexcrichton) [inline crates/wasi/src/preview2/ctx.rs:147]: While you're here mind adding documentation to this method? I forget how `perms` relates to `file_perms` myself

File: crates/wasi/src/preview2/filesystem.rs [hunk 3]
```diff
@@ -109,14 +111,21 @@ pub struct Dir {
     pub dir: Arc<cap_std::fs::Dir>,
     pub perms: DirPerms,
     pub file_perms: FilePerms,
+    pub open_mode: FilePerms,
```
Comment (alexcrichton) [inline crates/wasi/src/preview2/filesystem.rs:114]: Mind adding some brief comments here about what these fields are? Three sets of permissions seems like a good threshold for needing disambiguation

**Review by alexcrichton — APPROVED** [review]
Seems reasonable to me but I'm finding it difficult to follow what each of the sets of permissions are so I'm not 100% sure. Could you expand a bit on how this fixes https://github.com/bytecodealliance/wasmtime/issues/7829? I'm having a tough time connecting permissions and how that'd fail something fd-based vs path-based.

Also https://github.com/bytecodealliance/wasmtime/issues/7829 mentions errno 63 and ENOSR but I suspect that was something Python-specific or something like that?

File: crates/wasi/src/preview2/ctx.rs (hunk 2 above)
Comment (pchickey) [reply crates/wasi/src/preview2/ctx.rs]: Yep improved and documented this :)

**Comment by brettcannon** [comment]
I just finished running CPython's test suite and this fixed the issue I found and didn't cause any new ones!

**Pull request merged**